File: template/cpp/math/subset.cpp
#include <algorithm>
#include <bit>
#include <cstdint>
namespace OI::Math {
// https://leetcode.cn/problems/maximum-rows-covered-by-columns
// Gosper's Hack : https://zhuanlan.zhihu.com/p/360512296
// https://programmingforinsomniacs.blogspot.com/2018/03/gospers-hack-explained.html
/*
int m = 1<<n
for (auto i{(1u << k) - 1}; i < m; i = next_k_subset(i)) {
    // Any i contains k ones
    ...
}
*/
#define next_k_subset(i)  (((i ^ (i + (i & -i))) >> (std::countr_zero(i & -i) + 2)) | (i + (i & -i)))
#define in(x, begin, end) ((begin) <= (x) && (x) < (end))

// 遍历子集(每一个集合都有k个元素)
template<typename F>
void for_each_k_subset(uint32_t S, uint32_t k, F f) {
    for (uint32_t s{(1u << k) - 1}; s <= S; s = next_k_subset(s)) {
        // Any s contains k ones
        f(s);
    }
}
// 遍历子集(每一个集合都有k个元素)
template<typename R, typename F, typename SF>
R for_each_k_subset(uint32_t S, uint32_t k, F f, SF sf, R init) {
    for (uint32_t s{(1u << k) - 1}; s <= S; s = next_k_subset(s)) {
        init = sf(std::move(init), f(s));
    }
    return init;
}
template<typename F>
void for_each_subset(uint32_t S, F f) {
    uint32_t s = 0;
    do {
        f(s);
        s = (s - 1) & S;
    } while (s != 0u);
}
template<typename R, typename F, typename SF>
R for_each_subset(uint32_t S, F f, SF sf, R init) {
    uint32_t s = 0;
    do {
        init = sf(std::move(init), f(s));
        s = (s - 1) & S;
    } while (s != 0u);
    return init;
}
template<typename F>
void for_each_nonempty_subset(uint32_t S, F f) {
    for (uint32_t s{S}; s != 0; s = (s - 1) & S) {
        f(s);
    }
}
template<typename R, typename F, typename SF>
R for_each_nonempty_subset(uint32_t S, F f, SF sf, R init) {
    for (uint32_t s{S}; s != 0; s = (s - 1) & S) {
        init = sf(std::move(init), f(s));
    }
    return init;
}
} // namespace OI::Math

File: template/python/subset.py
from typing import Generator


def next_k_subset(s: int) -> int:
    """下一个k子集(用于k_subset方法)
    https://leetcode.cn/problems/maximum-rows-covered-by-columns
    Gosper's Hack : https://zhuanlan.zhihu.com/p/360512296
    https://programmingforinsomniacs.blogspot.com/2018/03/gospers-hack-explained.html
    """
    lb = s & -s
    r = s + lb
    return ((r ^ s) // lb >> 2) | r


def k_subset(S: int, k: int) -> Generator[int, None, None]:
    """生成子集(每一个集合都有k个元素)"""
    s = (1 << k) - 1
    while s <= S:
        yield s
        s = next_k_subset(s)


def subset(S: int) -> Generator[int, None, None]:
    """生成子集(含空集)"""
    s = 0
    while True:
        yield s
        s = (s - 1) & S
        if s == S:
            break


def nonempty_subset(S: int) -> Generator[int, None, None]:
    """生成非空子集"""
    s = S
    while s != 0:
        yield s
        s = (s - 1) & S


# 折半搜索
# https://leetcode.cn/problems/partition-array-into-two-arrays-to-minimize-sum-difference/description/
# https://leetcode.cn/problems/closest-subsequence-sum/description/


def tail_zero_count(x: int) -> int:
    """计算x的尾部0的个数"""
    s = 0
    while x & 1 == 0:
        s += 1
        x >>= 1
    return s


def subsequence_list(arr: list, op, initial) -> list:
    """获得arr的子集运算结果表"""
    n = len(arr)
    S = 1 << n
    res = [initial] * S
    for s in range(1, S):
        res[s] = op(res[s & (s - 1)], arr[tail_zero_count(s)])
    return res


def subsequence_count_list(arr: list, op, initial) -> list[list]:
    """获得arr的子集运算结果表并按二进制1的个数分组"""
    n = len(arr)
    res = [[] for _ in range(n + 1)]
    res[0].append(initial)
    S = 1 << n
    temp = [initial] * S
    for s in range(1, S):
        temp[s] = op(temp[s & (s - 1)], arr[tail_zero_count(s)])
        res[s.bit_count()].append(temp[s])
    return res
